{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CYP metabolizing phenotypes\n",
    "\n",
    "\"In mammals, __Cytochrome P450 (CYP)__ enzymes (...) are responsible for the oxidative metabolism of many xenobiotics as well as organic endogenous compounds. In humans, 57 isoforms were identified which are classified based on sequence homology\" [\\(Redlich et al. 2008\\)](https://pubmed.ncbi.nlm.nih.gov/18828626/).\n",
    "\n",
    "What are the frequencies of the __star allele__ haplotypes, and the observed\n",
    "diplotypes and associated __metabolizing phenotypes__ of two isoforms -\n",
    "CYP2C19 chromosome 10 and CYP2D6 chromosome 22 - in 1000 Genomes super-populations,\n",
    "and UKB imputed and whole exome sequencing data?\n",
    "\n",
    "<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sqlite3\n",
    "import vcf\n",
    "import subprocess\n",
    "import urllib.request\n",
    "import json\n",
    "\n",
    "from pathlib import Path\n",
    "from itertools import compress, combinations, combinations_with_replacement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "## Activity score, gene function, star allele, metabolizing phenotype\n",
    "\n",
    "From this [meta-analysis](https://www.nature.com/articles/s41398-020-01129-1):\n",
    "\n",
    "The __activity score (AS)__ system can be applied to both genes.\n",
    "\n",
    "| AS | function |\n",
    "|---|---|\n",
    "| 0 | non-functional |\n",
    "| 0.25 | decreased function |\n",
    "| 0.5 | decreased function |\n",
    "| 1 | normal function |\n",
    "| 1.5 | increased function |\n",
    "\n",
    "<br>\n",
    "\n",
    "Corresponding __star alleles__\n",
    "\n",
    "| AS | CYP2D6 | CYP2C19 |\n",
    "|---|---|---|\n",
    "| 0 | *3–*8, *15, *18, *31, *36, *47, *51, *56, *57, *62, *92, *100, *101 | *2, *3, *4, *5, *6, *7, *8, *23, *24 |\n",
    "| 0.25–0.5 | *9, *10, *17, *29, *41, *49, *50, *54, *55, *59, *72 | *9, *10, *12, *16, *25, *27 |\n",
    "| 1 | *1, *2, *27, *39, *45, *46, *48 | *1, *13, *15 *18 |\n",
    "| 1.5 | *53 | *17 |\n",
    "| Unknown |*43, *60, *65, *82, *84, *85, *86 | |\n",
    "\n",
    "<br>\n",
    "\n",
    "Associated diplotype __phenotypes__\n",
    "\n",
    "PM = poor metabolizer, IM = intermediate metabolizer, NM = normal metabolizer,\n",
    "RM = rapid metabolizer, UM = ultra-rapid metabolizer\n",
    "\n",
    "<br>\n",
    "\n",
    "| AS | phenotype | alleles |\n",
    "|---|---|---|\n",
    "| 0 | PM | homozygous carriers of non-functional alleles |\n",
    "| 0.25-1 | IM | one functional or decreased function allele and one non-functional allele and those carrying two decreased function alleles |\n",
    "| 1.25-2.25 | NM | homozygous carriers of normal function alleles and heterozygous carriers with one decreased function and one normal function allele |\n",
    "| >2.25 | UM | one or more increased function alleles and carriers of a duplication or multiplication of a functional allele |\n",
    "\n",
    "__Note.__ see the CPIC [Term Standardization for Clinical Pharmacogenetic Test Results Project](https://cpicpgx.org/resources/term-standardization/) and associated [publication](https://pubmed.ncbi.nlm.nih.gov/27441996/)\n",
    "\n",
    "<br>\n",
    "\n",
    "From [Dean 2017](https://www.ncbi.nlm.nih.gov/books/NBK425165/)\n",
    "\n",
    "__CYP2C19 metabolizer phenotypes__\n",
    "\n",
    "phenotype | e.g. diplotype | alleles |\n",
    "|---|---|---|\n",
    "| PM | *2/*2, *2/*3, *3/*3 | two non-functional alleles |\n",
    "| IM | *1/*2, *1/*3, *2/*17? | one normal function and one no function allele or one no function allele and one increased function allele |\n",
    "| NM | *1/*1 | two normal function alleles |\n",
    "| RM | *1/*17 | one normal function allele and one increased function allele |\n",
    "| UM | *17/*17 | two increased function alleles |\n",
    "\n",
    "\n",
    "__Note__. There are $k$-$combination + n$ (i.e. $n$ choose $k$ _with repitition_)\n",
    "possible diplotypes, where $k=2$ and $n$ is the number of star alleles.\n",
    "\n",
    "<br>\n",
    "\n",
    "## Variants used to call star alleles and phenotype\n",
    "\n",
    "From PharmGKB [CYP2C19](https://www.pharmgkb.org/page/cyp2c19RefMaterials) and\n",
    "[CYP2D6](https://www.pharmgkb.org/page/cyp2d6RefMaterials) reference material.\n",
    "\n",
    "<br>\n",
    "\n",
    "### CYP2C19 - chromosome 10\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def display_df(data, nrow=6):\n",
    "    \"\"\"Display head and tail of dataframe\"\"\"\n",
    "    with pd.option_context('display.max_rows', nrow):\n",
    "        display(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CPIC CYP2C19 variant-star allele definition table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/k1342162/apps/miniconda3/lib/python3.7/site-packages/openpyxl/styles/stylesheet.py:226: UserWarning: Workbook contains no default style, apply openpyxl's default\n",
      "  warn(\"Workbook contains no default style, apply openpyxl's default\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NG_008384.3 (ATG start)</th>\n",
       "      <th>Effect on protein (NP_000760.1)</th>\n",
       "      <th>Position at NC_000010.11 (Homo sapiens chromosome 10, GRCh38.p13)</th>\n",
       "      <th>Position at NG_008384.3 (CYP2C19 RefSeqGene; forward relative to chromosome)</th>\n",
       "      <th>rsID</th>\n",
       "      <th>CYP2C19 Allele</th>\n",
       "      <th>*38</th>\n",
       "      <th>*1</th>\n",
       "      <th>*2</th>\n",
       "      <th>*3</th>\n",
       "      <th>...</th>\n",
       "      <th>*29</th>\n",
       "      <th>*30</th>\n",
       "      <th>*31</th>\n",
       "      <th>*32</th>\n",
       "      <th>*33</th>\n",
       "      <th>*34</th>\n",
       "      <th>*35</th>\n",
       "      <th>*36</th>\n",
       "      <th>*37</th>\n",
       "      <th>*39</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-806C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.94761900C&gt;T</td>\n",
       "      <td>g.4220C&gt;T</td>\n",
       "      <td>rs12248560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1A&gt;G</td>\n",
       "      <td>p.M1V</td>\n",
       "      <td>g.94762706A&gt;G</td>\n",
       "      <td>g.5026A&gt;G</td>\n",
       "      <td>rs28399504</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7C&gt;T</td>\n",
       "      <td>p.P3S</td>\n",
       "      <td>g.94762712C&gt;T</td>\n",
       "      <td>g.5032C&gt;T</td>\n",
       "      <td>rs367543002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>90080C&gt;G</td>\n",
       "      <td>p.F448L</td>\n",
       "      <td>g.94852785C&gt;G</td>\n",
       "      <td>g.95105C&gt;G</td>\n",
       "      <td>rs118203759</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>90209A&gt;C</td>\n",
       "      <td>p.X491C</td>\n",
       "      <td>g.94852914A&gt;C</td>\n",
       "      <td>g.95234A&gt;C</td>\n",
       "      <td>rs55640102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Structural Variation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CYP2C19 full gene deletions</td>\n",
       "      <td>CYP2C19 partial gene deletions</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   NG_008384.3 (ATG start) Effect on protein (NP_000760.1)  \\\n",
       "1                  -806C>T                             NaN   \n",
       "2                     1A>G                           p.M1V   \n",
       "3                     7C>T                           p.P3S   \n",
       "..                     ...                             ...   \n",
       "33                90080C>G                         p.F448L   \n",
       "34                90209A>C                         p.X491C   \n",
       "35    Structural Variation                             NaN   \n",
       "\n",
       "   Position at NC_000010.11 (Homo sapiens chromosome 10, GRCh38.p13)  \\\n",
       "1                                       g.94761900C>T                  \n",
       "2                                       g.94762706A>G                  \n",
       "3                                       g.94762712C>T                  \n",
       "..                                                ...                  \n",
       "33                                      g.94852785C>G                  \n",
       "34                                      g.94852914A>C                  \n",
       "35                                                NaN                  \n",
       "\n",
       "   Position at NG_008384.3 (CYP2C19 RefSeqGene; forward relative to chromosome)  \\\n",
       "1                                           g.4220C>T                             \n",
       "2                                           g.5026A>G                             \n",
       "3                                           g.5032C>T                             \n",
       "..                                                ...                             \n",
       "33                                         g.95105C>G                             \n",
       "34                                         g.95234A>C                             \n",
       "35                                                NaN                             \n",
       "\n",
       "           rsID CYP2C19 Allele  *38   *1   *2   *3  ...  *29  *30  *31  *32  \\\n",
       "1    rs12248560            NaN    C  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "2    rs28399504            NaN    A  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "3   rs367543002            NaN    C  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "..          ...            ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "33  rs118203759            NaN    C  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "34   rs55640102            NaN    A  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "35          NaN            NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "\n",
       "    *33  *34  *35                          *36  \\\n",
       "1   NaN  NaN  NaN                          NaN   \n",
       "2   NaN  NaN  NaN                          NaN   \n",
       "3   NaN    T  NaN                          NaN   \n",
       "..  ...  ...  ...                          ...   \n",
       "33  NaN  NaN  NaN                          NaN   \n",
       "34  NaN  NaN  NaN                          NaN   \n",
       "35  NaN  NaN  NaN  CYP2C19 full gene deletions   \n",
       "\n",
       "                               *37  *39  \n",
       "1                              NaN  NaN  \n",
       "2                              NaN  NaN  \n",
       "3                              NaN  NaN  \n",
       "..                             ...  ...  \n",
       "33                             NaN  NaN  \n",
       "34                             NaN  NaN  \n",
       "35  CYP2C19 partial gene deletions  NaN  \n",
       "\n",
       "[35 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cyp2c19_variants = pd.read_excel('cpic/CYP2C19_allele_definition_table.xlsx', skiprows=1, header=None).transpose()\n",
    "cyp2c19_variants = cyp2c19_variants.rename(columns=cyp2c19_variants.iloc[0]).drop(cyp2c19_variants.index[0])\n",
    "\n",
    "display_df(cyp2c19_variants)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "CPIC CYP2C19 allele functionality table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele/cDNA/rsID</th>\n",
       "      <th>Allele Clinical Functional Status (Required)</th>\n",
       "      <th>PMID (Required)</th>\n",
       "      <th>Strength of Evidence  (Required)</th>\n",
       "      <th>Summary of Findings (Required)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*1</td>\n",
       "      <td>Normal function</td>\n",
       "      <td>7487078, 32602114, 22027650</td>\n",
       "      <td>Definitve</td>\n",
       "      <td>CYP2C19*1 is assigned normal function based on...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*2</td>\n",
       "      <td>No function</td>\n",
       "      <td>8195181, 22027650</td>\n",
       "      <td>Definitve</td>\n",
       "      <td>CYP2C19*2 is assigned no function based on def...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>*3</td>\n",
       "      <td>No function</td>\n",
       "      <td>7969038, 9103550, 22027650</td>\n",
       "      <td>Definitve</td>\n",
       "      <td>CYP2C19*3 is assigned no function based on def...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>*36</td>\n",
       "      <td>No function</td>\n",
       "      <td>31260137</td>\n",
       "      <td>Limited</td>\n",
       "      <td>CYP2C19*36 is assigned no function based on li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>*37</td>\n",
       "      <td>No function</td>\n",
       "      <td>31260137</td>\n",
       "      <td>Limited</td>\n",
       "      <td>CYP2C19*37 is assigned no function based on li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>*38</td>\n",
       "      <td>Normal function</td>\n",
       "      <td>2009263, 32602114</td>\n",
       "      <td>Definitve</td>\n",
       "      <td>CYP2C19*38 is assigned normal function based o...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Allele/cDNA/rsID Allele Clinical Functional Status (Required)  \\\n",
       "0                *1                              Normal function   \n",
       "1                *2                                  No function   \n",
       "2                *3                                  No function   \n",
       "..              ...                                          ...   \n",
       "32              *36                                  No function   \n",
       "33              *37                                  No function   \n",
       "34              *38                              Normal function   \n",
       "\n",
       "                PMID (Required) Strength of Evidence  (Required)  \\\n",
       "0   7487078, 32602114, 22027650                        Definitve   \n",
       "1             8195181, 22027650                        Definitve   \n",
       "2    7969038, 9103550, 22027650                        Definitve   \n",
       "..                          ...                              ...   \n",
       "32                     31260137                          Limited   \n",
       "33                     31260137                          Limited   \n",
       "34            2009263, 32602114                        Definitve   \n",
       "\n",
       "                       Summary of Findings (Required)  \n",
       "0   CYP2C19*1 is assigned normal function based on...  \n",
       "1   CYP2C19*2 is assigned no function based on def...  \n",
       "2   CYP2C19*3 is assigned no function based on def...  \n",
       "..                                                ...  \n",
       "32  CYP2C19*36 is assigned no function based on li...  \n",
       "33  CYP2C19*37 is assigned no function based on li...  \n",
       "34  CYP2C19*38 is assigned normal function based o...  \n",
       "\n",
       "[35 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cyp2c19_function = (pd.read_excel('cpic/CYP2C19_allele_functionality_reference.xlsx', skiprows=1)\n",
    "                    .dropna(axis='columns', how='all'))\n",
    "display_df(cyp2c19_function)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "No function           9\n",
       "Uncertain function    9\n",
       "Normal function       7\n",
       "Decreased function    6\n",
       " No function          3\n",
       "Increased function    1\n",
       "Name: Allele Clinical Functional Status (Required), dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cyp2c19_function['Allele Clinical Functional Status (Required)'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele/cDNA/rsID</th>\n",
       "      <th>Allele Clinical Functional Status (Required)</th>\n",
       "      <th>PMID (Required)</th>\n",
       "      <th>Strength of Evidence  (Required)</th>\n",
       "      <th>Summary of Findings (Required)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>*17</td>\n",
       "      <td>Increased function</td>\n",
       "      <td>16413245, 17625515, 20083681</td>\n",
       "      <td>Strong</td>\n",
       "      <td>CYP2C19*17 is assigned increased function base...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Allele/cDNA/rsID Allele Clinical Functional Status (Required)  \\\n",
       "16              *17                           Increased function   \n",
       "\n",
       "                 PMID (Required) Strength of Evidence  (Required)  \\\n",
       "16  16413245, 17625515, 20083681                           Strong   \n",
       "\n",
       "                       Summary of Findings (Required)  \n",
       "16  CYP2C19*17 is assigned increased function base...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cyp2c19_function[cyp2c19_function['Allele/cDNA/rsID'].isin(['*17'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "CPIC CYP2C19 diplotype-phenotype table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/k1342162/apps/miniconda3/lib/python3.7/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CYP2C19 Diplotype</th>\n",
       "      <th>CYP2C19 phenotype</th>\n",
       "      <th>EHR Priority Result Notation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*1/*2</td>\n",
       "      <td>CYP2C19 Intermediate Metabolizer</td>\n",
       "      <td>Abnormal/Priority/High Risk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*1/*22</td>\n",
       "      <td>CYP2C19 Intermediate Metabolizer</td>\n",
       "      <td>Abnormal/Priority/High Risk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>*1/*24</td>\n",
       "      <td>CYP2C19 Intermediate Metabolizer</td>\n",
       "      <td>Abnormal/Priority/High Risk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>*9/*32</td>\n",
       "      <td>Indeterminate</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>*9/*33</td>\n",
       "      <td>Indeterminate</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>*9/*34</td>\n",
       "      <td>Indeterminate</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>630 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    CYP2C19 Diplotype                  CYP2C19 phenotype  \\\n",
       "0               *1/*2  CYP2C19 Intermediate Metabolizer    \n",
       "1              *1/*22  CYP2C19 Intermediate Metabolizer    \n",
       "2              *1/*24  CYP2C19 Intermediate Metabolizer    \n",
       "..                ...                                ...   \n",
       "627            *9/*32                      Indeterminate   \n",
       "628            *9/*33                      Indeterminate   \n",
       "629            *9/*34                      Indeterminate   \n",
       "\n",
       "    EHR Priority Result Notation  \n",
       "0    Abnormal/Priority/High Risk  \n",
       "1    Abnormal/Priority/High Risk  \n",
       "2    Abnormal/Priority/High Risk  \n",
       "..                           ...  \n",
       "627                         none  \n",
       "628                         none  \n",
       "629                         none  \n",
       "\n",
       "[630 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "cyp2c19_diplotype = (pd.read_excel('cpic/CYP2C19_Diplotype_Phenotype_Table.xlsx', skipfooter=4)\n",
    "                     .dropna(axis='columns', how='all'))\n",
    "display_df(cyp2c19_diplotype)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not all mathematically possible combinations appear the CPIC diplotype table, e.g.,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "150 possible CYP2C19 diplotypes not included in CPIC table\n"
     ]
    }
   ],
   "source": [
    "possible_combinations = list(combinations_with_replacement(['*' + str(x) for x in list(range(1, 39 +1))], 2))\n",
    "possible_combinations = ['/'.join(x) for x in possible_combinations]\n",
    "\n",
    "print(\n",
    "    len(list(set(possible_combinations) - set(cyp2c19_diplotype['CYP2C19 Diplotype']))),\n",
    "    'possible CYP2C19 diplotypes not included in CPIC table')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsID</th>\n",
       "      <th>Position at NC_000010.11 (Homo sapiens chromosome 10, GRCh38.p13)</th>\n",
       "      <th>*17</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs12248560</td>\n",
       "      <td>g.94761900C&gt;T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rs3758581</td>\n",
       "      <td>g.94842866A&gt;G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          rsID  \\\n",
       "1   rs12248560   \n",
       "27   rs3758581   \n",
       "\n",
       "   Position at NC_000010.11 (Homo sapiens chromosome 10, GRCh38.p13) *17  \n",
       "1                                       g.94761900C>T                  T  \n",
       "27                                      g.94842866A>G                  G  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Star alleles corresponding to metabolizing phenotypes\n",
    "\n",
    "# ['*2', '*3', '*4', '*5', '*6', '*7', '*8', '*23', '*24']\n",
    "# ['*9', '*10', '*12', '*16', '*25', '*27']\n",
    "# cyp2c19_variants[['rsID', '*1', '*13', '*15', '*18']].dropna(how='all', subset=['*1', '*13', '*15', '*18'])\n",
    "\n",
    "# Increased function\n",
    "cyp2c19_variants[['rsID', 'Position at NC_000010.11 (Homo sapiens chromosome 10, GRCh38.p13)', '*17']].dropna(how='all', subset=['*17'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "### CYP2D6 - chromosome 22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/k1342162/apps/miniconda3/lib/python3.7/site-packages/openpyxl/styles/stylesheet.py:226: UserWarning: Workbook contains no default style, apply openpyxl's default\n",
      "  warn(\"Workbook contains no default style, apply openpyxl's default\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NG_008376.4 (ATG start)</th>\n",
       "      <th>Effect on protein (NP_000097.3)</th>\n",
       "      <th>Position at NC_000022.11 (Homo sapiens chromosome 22, GRCh38.p13)</th>\n",
       "      <th>Position at NG_008376.4 (CYP2D6 RefSeqGene; reverse relative to chromosome)</th>\n",
       "      <th>rsID</th>\n",
       "      <th>CYP2D6 Allele</th>\n",
       "      <th>*1</th>\n",
       "      <th>*2</th>\n",
       "      <th>*3</th>\n",
       "      <th>*4</th>\n",
       "      <th>...</th>\n",
       "      <th>*139</th>\n",
       "      <th>*140</th>\n",
       "      <th>*141</th>\n",
       "      <th>*142</th>\n",
       "      <th>*143</th>\n",
       "      <th>*144</th>\n",
       "      <th>*145</th>\n",
       "      <th>*146</th>\n",
       "      <th>*147</th>\n",
       "      <th>*149</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14C&gt;T</td>\n",
       "      <td>p.A5V</td>\n",
       "      <td>g.42130778G&gt;A</td>\n",
       "      <td>g.5033C&gt;T</td>\n",
       "      <td>rs773790593</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19G&gt;A</td>\n",
       "      <td>p.V7M</td>\n",
       "      <td>g.42130773C&gt;T</td>\n",
       "      <td>g.5038G&gt;A</td>\n",
       "      <td>rs72549358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>31G&gt;A</td>\n",
       "      <td>p.V11M</td>\n",
       "      <td>g.42130761C&gt;T</td>\n",
       "      <td>g.5050G&gt;A</td>\n",
       "      <td>rs769258</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>4187C&gt;T</td>\n",
       "      <td>p.S488F</td>\n",
       "      <td>g.42126605G&gt;A</td>\n",
       "      <td>g.9206C&gt;T</td>\n",
       "      <td>rs568495591</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>4214G&gt;A</td>\n",
       "      <td>p.R497H</td>\n",
       "      <td>g.42126578C&gt;T</td>\n",
       "      <td>g.9233G&gt;A</td>\n",
       "      <td>rs1440526469</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Structural Variation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows × 146 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    NG_008376.4 (ATG start) Effect on protein (NP_000097.3)  \\\n",
       "1                     14C>T                           p.A5V   \n",
       "2                     19G>A                           p.V7M   \n",
       "3                     31G>A                          p.V11M   \n",
       "..                      ...                             ...   \n",
       "126                 4187C>T                         p.S488F   \n",
       "127                 4214G>A                         p.R497H   \n",
       "128    Structural Variation                             NaN   \n",
       "\n",
       "    Position at NC_000022.11 (Homo sapiens chromosome 22, GRCh38.p13)  \\\n",
       "1                                        g.42130778G>A                  \n",
       "2                                        g.42130773C>T                  \n",
       "3                                        g.42130761C>T                  \n",
       "..                                                 ...                  \n",
       "126                                      g.42126605G>A                  \n",
       "127                                      g.42126578C>T                  \n",
       "128                                                NaN                  \n",
       "\n",
       "    Position at NG_008376.4 (CYP2D6 RefSeqGene; reverse relative to chromosome)  \\\n",
       "1                                            g.5033C>T                            \n",
       "2                                            g.5038G>A                            \n",
       "3                                            g.5050G>A                            \n",
       "..                                                 ...                            \n",
       "126                                          g.9206C>T                            \n",
       "127                                          g.9233G>A                            \n",
       "128                                                NaN                            \n",
       "\n",
       "             rsID CYP2D6 Allele   *1   *2   *3   *4  ... *139 *140 *141 *142  \\\n",
       "1     rs773790593           NaN    G  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "2      rs72549358           NaN    C  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "3        rs769258           NaN    C  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "..            ...           ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "126   rs568495591           NaN    G  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "127  rs1440526469           NaN    C  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "128           NaN           NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN   \n",
       "\n",
       "    *143 *144 *145 *146 *147 *149  \n",
       "1    NaN  NaN  NaN  NaN  NaN  NaN  \n",
       "2    NaN  NaN  NaN  NaN  NaN  NaN  \n",
       "3      T  NaN  NaN  NaN  NaN  NaN  \n",
       "..   ...  ...  ...  ...  ...  ...  \n",
       "126  NaN  NaN  NaN  NaN  NaN  NaN  \n",
       "127  NaN  NaN  NaN  NaN  NaN  NaN  \n",
       "128  NaN  NaN  NaN  NaN  NaN  NaN  \n",
       "\n",
       "[128 rows x 146 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cyp2d6_variants = pd.read_excel('cpic/CYP2D6_allele_definition_table.xlsx', skiprows=1, header=None).transpose()\n",
    "cyp2d6_variants = cyp2d6_variants.rename(columns=cyp2d6_variants.iloc[0]).drop(cyp2d6_variants.index[0])\n",
    "display_df(cyp2d6_variants)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CYP2D6 diplotype-phenotype table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CYP2D6 Diplotype</th>\n",
       "      <th>Activity Score</th>\n",
       "      <th>Coded Diplotype/Phenotype Summary</th>\n",
       "      <th>EHR Priority Notation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*1/*1</td>\n",
       "      <td>2</td>\n",
       "      <td>Normal Metabolizer</td>\n",
       "      <td>Normal/Routine/ Low Risk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*1/*1x2</td>\n",
       "      <td>3</td>\n",
       "      <td>Ultrarapid Metabolizer</td>\n",
       "      <td>Abnormal/Priority/High Risk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>*1/*1≥3</td>\n",
       "      <td>≥4</td>\n",
       "      <td>Ultrarapid Metabolizer</td>\n",
       "      <td>Abnormal/Priority/High Risk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11023</th>\n",
       "      <td>*138/*138</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Indeterminate</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11024</th>\n",
       "      <td>*138/*139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Indeterminate</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11025</th>\n",
       "      <td>*139/*139</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Indeterminate</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11026 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CYP2D6 Diplotype Activity Score Coded Diplotype/Phenotype Summary  \\\n",
       "0                *1/*1              2                Normal Metabolizer   \n",
       "1              *1/*1x2              3            Ultrarapid Metabolizer   \n",
       "2              *1/*1≥3             ≥4            Ultrarapid Metabolizer   \n",
       "...                ...            ...                               ...   \n",
       "11023        *138/*138            NaN                     Indeterminate   \n",
       "11024        *138/*139            NaN                     Indeterminate   \n",
       "11025        *139/*139            NaN                     Indeterminate   \n",
       "\n",
       "             EHR Priority Notation  \n",
       "0         Normal/Routine/ Low Risk  \n",
       "1      Abnormal/Priority/High Risk  \n",
       "2      Abnormal/Priority/High Risk  \n",
       "...                            ...  \n",
       "11023                         none  \n",
       "11024                         none  \n",
       "11025                         none  \n",
       "\n",
       "[11026 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cyp2d6_diplotype = (pd.read_excel('cpic/CYP2D6_Diplotype_Phenotype_Table.xlsx')\n",
    "                     .dropna(axis='columns', how='all'))\n",
    "display_df(cyp2d6_diplotype)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "CPIC allele functionality table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Activity Value (Optional)</th>\n",
       "      <th>Allele Clinical Function Status (Required)*</th>\n",
       "      <th>PMID (Optional)</th>\n",
       "      <th>Strength of Evidence (Optional)</th>\n",
       "      <th>Findings (Optional)</th>\n",
       "      <th>Comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Normal function</td>\n",
       "      <td>2574001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2574001: debrisoquine (in vivo)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*1x2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Increased function</td>\n",
       "      <td>7616439; 9012401; 17259947; 19541866; 22111604...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7616439: debrisoquine (in vivo); 9012401: dext...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>*1≥3</td>\n",
       "      <td>≥3.0</td>\n",
       "      <td>Increased function</td>\n",
       "      <td>7616439; 9012401; 17259947; 19541866; 22111604...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>152 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Activity Value (Optional)  \\\n",
       "0      *1                       1.0   \n",
       "1    *1x2                       2.0   \n",
       "2    *1≥3                      ≥3.0   \n",
       "..    ...                       ...   \n",
       "149   NaN                       NaN   \n",
       "150   NaN                       NaN   \n",
       "151   NaN                       NaN   \n",
       "\n",
       "    Allele Clinical Function Status (Required)*  \\\n",
       "0                               Normal function   \n",
       "1                           Increased function    \n",
       "2                           Increased function    \n",
       "..                                          ...   \n",
       "149                                         NaN   \n",
       "150                                         NaN   \n",
       "151                                         NaN   \n",
       "\n",
       "                                       PMID (Optional)  \\\n",
       "0                                              2574001   \n",
       "1    7616439; 9012401; 17259947; 19541866; 22111604...   \n",
       "2    7616439; 9012401; 17259947; 19541866; 22111604...   \n",
       "..                                                 ...   \n",
       "149                                                NaN   \n",
       "150                                                NaN   \n",
       "151                                                NaN   \n",
       "\n",
       "    Strength of Evidence (Optional)  \\\n",
       "0                               NaN   \n",
       "1                               NaN   \n",
       "2                               NaN   \n",
       "..                              ...   \n",
       "149                             NaN   \n",
       "150                             NaN   \n",
       "151                             NaN   \n",
       "\n",
       "                                   Findings (Optional) Comments  \n",
       "0                      2574001: debrisoquine (in vivo)      NaN  \n",
       "1    7616439: debrisoquine (in vivo); 9012401: dext...      NaN  \n",
       "2                                                  NaN      NaN  \n",
       "..                                                 ...      ...  \n",
       "149                                                NaN      NaN  \n",
       "150                                                NaN      NaN  \n",
       "151                                                NaN      NaN  \n",
       "\n",
       "[152 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cyp2d6_function = (pd.read_excel('cpic/CYP2D6_allele_functionality_reference.xlsx', skiprows=1)\n",
    "                    .dropna(axis='columns', how='all'))\n",
    "display_df(cyp2d6_function)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Activity Value (Optional)</th>\n",
       "      <th>Allele Clinical Function Status (Required)*</th>\n",
       "      <th>PMID (Optional)</th>\n",
       "      <th>Strength of Evidence (Optional)</th>\n",
       "      <th>Findings (Optional)</th>\n",
       "      <th>Comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>*5</td>\n",
       "      <td>0</td>\n",
       "      <td>No function</td>\n",
       "      <td>1673290; 11266079</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1673290: debrisoquine, sparteine (in vivo); 11...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Activity Value (Optional) Allele Clinical Function Status (Required)*  \\\n",
       "10  *5                         0                                 No function   \n",
       "\n",
       "      PMID (Optional) Strength of Evidence (Optional)  \\\n",
       "10  1673290; 11266079                             NaN   \n",
       "\n",
       "                                  Findings (Optional) Comments  \n",
       "10  1673290: debrisoquine, sparteine (in vivo); 11...      NaN  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 'Allele/cDNA/rsID'\n",
    "# cyp2d6_function.columns\n",
    "cyp2d6_function[cyp2d6_function['  '].isin(['*5'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variants in UKB imputed genetic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rsid_count(data, chromosome, gene):\n",
    "    \"\"\"Count variants in CPIC reference and imputed data\n",
    "\n",
    "    data (pd.DataFrame): transposed CPIC defintion table as dataframe\n",
    "    chromosome (int): chromosome number\n",
    "    gene (str): PGx gene\n",
    "    \"\"\"\n",
    "    rsids = list(compress(\n",
    "        data['rsID'].dropna(),\n",
    "        data['rsID'].dropna().str.startswith('rs')\n",
    "        ))\n",
    "    \n",
    "    print(len(rsids), f'rsID SNPs used to call {gene} star alleles')\n",
    "    \n",
    "    bgen_file_list = [\n",
    "        f'/scratch/datasets/ukbiobank/June2017/Imputed/ukb_imp_chr{chromosome}_v3_MAF1_INFO4.bgen.bgi',\n",
    "        f'/scratch/datasets/ukbiobank/June2017/Imputed/ukb_imp_chr{chromosome}_v3_MAF0_INFO7.bgen.bgi',\n",
    "        f'/scratch/datasets/ukbiobank/June2017/Imputed/ukb_imp_chr{chromosome}_v3.bgen.bgi']\n",
    "    \n",
    "    for f in bgen_file_list:\n",
    "        file_name = Path(f).name\n",
    "        con = sqlite3.connect(f)\n",
    "        bgi = pd.read_sql_query('SELECT * FROM Variant', con)\n",
    "        print(bgi[bgi['rsid'].isin(rsids)].shape[0], f'of {len(rsids)} included in {file_name}')\n",
    "        con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34 rsID SNPs used to call CYP2C19 star alleles\n",
      "4 of 34 included in ukb_imp_chr10_v3_MAF1_INFO4.bgen.bgi\n",
      "16 of 34 included in ukb_imp_chr10_v3_MAF0_INFO7.bgen.bgi\n",
      "19 of 34 included in ukb_imp_chr10_v3.bgen.bgi\n"
     ]
    }
   ],
   "source": [
    "rsid_count(cyp2c19_variants, 10, 'CYP2C19')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally full deletion corresponds to *36; partial deletion to *37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cyp2c19_variants.loc[cyp2c19_variants['rsID'].isna(), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "Chromosome 22 variants used to call CYP2D6 star alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid_count(cyp2d6_variants, 22, 'CYP2D6')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cyp2d6_variants.loc[cyp2d6_variants['rsID'].isna(), ].dropna(axis='columns', how='all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "Write to file the list of rsID variants observed in the unfiltered UKB data. Note. this does not includes indels - check how these are encoded captured in the VCF file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def write_rsids(csv_file, data, chromosome):\n",
    "    \"\"\"Write to file the rsIDs observed in the unfiltered UKB imputed data\"\"\"\n",
    "    rsids = list(compress(\n",
    "        data['rsID'].dropna(),\n",
    "        data['rsID'].dropna().str.startswith('rs')\n",
    "        ))\n",
    "    \n",
    "    path = f'/scratch/datasets/ukbiobank/June2017/Imputed/ukb_imp_chr{chromosome}_v3.bgen.bgi'\n",
    "    con = sqlite3.connect(path)\n",
    "    bgi = pd.read_sql_query('SELECT * FROM Variant', con)\n",
    "    bgi[bgi['rsid'].isin(rsids)]['rsid'].to_csv(csv_file, header=False, index=False)\n",
    "    con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "write_rsids('cyp2c19_rsids.csv', cyp2c19_variants, 10)\n",
    "write_rsids('cyp2d6_rsids.csv', cyp2d6_variants, 22)\n",
    "\n",
    "# Full list from PharmGKB\n",
    "# cyp2c19_variants['rsID'].dropna().to_csv('cyp2c19_rsids.csv', header=False, index=False)\n",
    "# cyp2d6_variants['rsID'].dropna().to_csv('cyp2d6_rsids.csv', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.3 64-bit ('base': conda)",
   "language": "python",
   "name": "python373jvsc74a57bd022ab3e6a73a2731b2d2a5eee53baffdb723c6979335821acd0eb70065f9061b3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
